✨ Show label when live story is updated and user is on last page.

Enriqe opened the pull request:
Follow up for #22519
Partial for #21714

Shows a label when a live story is updated with new pages and the user is on the last page.

Preview:


![ezgif-1-0464c6e653dd](https://user-images.githubusercontent.com/5449100/58978058-0aa76780-8799-11e9-9b44-05446c9e0a0e.gif)

EDIT: updated UX
![liveupdate](https://user-images.githubusercontent.com/5449100/59309882-c45b7800-8c72-11e9-895c-7407cd43c220.gif)

![3panelsupdatev2](https://user-images.githubusercontent.com/5449100/59310067-4055c000-8c73-11e9-8273-c3f5b1b09ec3.gif)

gmajoulet commented on extensions/amp-story/1.0/_locales/en.js:
```diff
@@ -98,6 +98,11 @@ const strings = {
+  [LocalizedStringId.AMP_STORY_HAS_NEW_PAGE_TEXT]: {
+    string: 'Story updated',
+    description:
+      'Label indicating that the story has been updated with (a) new page(s).',
```
Since the translators have 0 context about the product, should we make this description less specific, and maybe say that new content is available?

newmuis replied:
Maybe let's try "Label that indicates that additional content has been added to a story"

WDYT?

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.js:
```diff
@@ -88,6 +84,13 @@ const TEMPLATE = {
       tag: 'div',
       attrs: dict({'class': 'i-amphtml-story-system-layer-buttons'}),
       children: [
+        {
```
This would break for a few cases:
- RTL
- First page where the (i) icon is visible

Can we use the same message container that's already implemented, instead of adding a new one? Making it generic looks trivial, I think renaming `.i-amphtml-story-sound-display` should be enough. Then you'd get RTL and (i) icon support out of the box.

newmuis replied:
Is the UX the same?  I think the audio message "comes out of" the mute/unmute icon.  Is that the same for this?

Enriqe replied:
Asking @hongwei1990 , I'll update the PR when I get the details.

gmajoulet replied:
Technically the text is displayed by the buttons. The audio icon happens to be the first one, so we get that effect, but if there's the (i) icon we wouldn't
It's Hong's call, but from what I see in this PR it should be fine :)

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.js:
```diff
@@ -236,6 +232,9 @@ export class SystemLayer {
     this.timeoutId_ = null;
+
+    /** @private {?number|?string} */
+    this.storyHasNewPageTimeoutId_ = null;
```
I think we want to use the same `timeoutId_` for all the messages that can be displayed in the system layer, otherwise they'd just be displayed on top of each other (and that'll make the code easier).

Enriqe commented:
Updated the PR with the most recent UX spec. PTAL.

See previews below for reference:
![mobileupdate](https://user-images.githubusercontent.com/5449100/59224989-a96b0400-8b9d-11e9-83c3-52febf9a5c93.gif)
![3panelsupdate](https://user-images.githubusercontent.com/5449100/59224990-a96b0400-8b9d-11e9-88c4-4be1c52f38e3.gif)

You'll notice that on the three panel view, the pagination button isn't updated when there is a new page. This is tracked here #22770

newmuis commented:
| Actual | Expected |
|:------:|:----------:|
| ![Screenshot of actual live list indicator UI](https://user-images.githubusercontent.com/1651960/59226753-ba1d7900-8ba1-11e9-9f19-09cbfa011d12.png) | ![Mock of expected live list indicator UI](https://user-images.githubusercontent.com/1651960/59226912-0e285d80-8ba2-11e9-9f5f-2a34480d3b10.png) |

Some quick minor UI nits:

- Can't tell if it's just because of the low fidelity of the GIF, but it looks like there's only one circle and not two?
- The font should be bold
- The mocks look like the corners are not completely sharp; @hongwei1990 can you verify?
- It looks like there's too much space on the left and/or too little space on the right

Note that I haven't actually taken a look at the code yet; this is all based on the rendering in the GIF

Enriqe commented:
> Can't tell if it's just because of the low fidelity of the GIF, but it looks like there's only one circle and not two?

Yep, it's the gif, sorry. Here is a screenshot with a better resolution of how it actually looks. You can see that the font is already bold.

I corrected the opacity of the outer circle to the spec and now it's clearer that there are two circles. I also added a border radius for the box edges but I will increase it to make it look more like the mocks since I didn't get an exact number.

Also added a bit more space on the right.

![image](https://user-images.githubusercontent.com/5449100/59228523-0ec2f300-8ba6-11e9-80c3-dc992943cfef.png)

newmuis commented on extensions/amp-story/1.0/amp-story-system-layer.js:
```diff
@@ -548,29 +553,31 @@ export class SystemLayer {
     }
     this.vsync_.mutate(() => {
-      this.getShadowRoot().setAttribute(MESSAGE_DISPLAY_CLASS, 'noshow');
+      this.getShadowRoot().setAttribute(message, 'noshow');
```
Sorry I know this is the type of thing that's very difficult to test (manually, or in any automated fashion), but how do multiple messages override each other?  Does the first one just disgust completely once the second one comes in?

I suspect that would be the case, but I'm not actually sure whether that's okay?  This can result in weird edge cases when one message is triggered right after another.  It's probably fine if it's two audio message since they'd both come from user interactions, but if it's an audio message followed by a live list update, the user won't have had time to read the audio message.

Should we consider queuing the messages like toast frameworks (likely) would?

Enriqe replied:
I tested it and right now they are both shown, and eventually disappear. Since this seems to be a very specific edge case, do you mind if I address it in a follow-up?

I filed #22778 to keep track.

gmajoulet commented:
Could you please post a screenshot of that UI when displayed RTL? The last
margin you added might break that display :(

On Mon, Jun 10, 2019 at 5:37 PM Enrique Marroquin <notifications@github.com>
wrote:

> Can't tell if it's just because of the low fidelity of the GIF, but it
> looks like there's only one circle and not two?
>
> Yep, it's the gif, sorry. Here is a screenshot with a better resolution of
> how it actually looks. You can see that the font is already bold.
>
> I corrected the opacity of the outer circle to the spec and now it's
> clearer that there are two circles. I also added a border radius for the
> box edges but I will increase it to make it look more like the mocks since
> I didn't get an exact number.
>
> Also added a bit more space on the right.
>
> [image: image]
> <https://user-images.githubusercontent.com/5449100/59228523-0ec2f300-8ba6-11e9-80c3-dc992943cfef.png>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/22700?email_source=notifications&email_token=AALMITDZH6UCF3XFO2ENJQLPZ3CRLA5CNFSM4HUBN3KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXLJ5FI#issuecomment-500604565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALMITGEHY45TRXV7FRLDXDPZ3CRLANCNFSM4HUBN3KA>
> .
>
-- 
Gabriel Majoulet  | UX Engineer | gmajoulet@google.com

newmuis commented on extensions/amp-story/1.0/amp-story-system-layer.js:
```diff
@@ -617,7 +624,13 @@ export class SystemLayer {
+      const lastIndex =
+        this.storeService_.get(StateProperty.PAGE_IDS).length - 1;
       this.getShadowRoot().classList.toggle('first-page-active', index === 0);
+      this.getShadowRoot().classList.toggle(
```
Can we please prefix these classes with `i-amphtml-`?

Enriqe replied:
Done.

newmuis commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -243,13 +252,48 @@
+
+.i-amphtml-story-has-new-page-circle-icon {
+	background: rgba(3, 255, 160, 1) !important;
+	border-radius: 50% !important;
```
This should have a `4px` border at `rgba(3, 255, 160, 0.5)`

Enriqe replied:
Did you mean to comment this on the line below of the box-shadow, which represents the outer circle?

Enriqe replied:
If so, according to the spec the outer circle is supposed to be 10px in width, while the inner circle is 6px in width.

And if I understand correctly, the box-shadow `spread` property specifies the *radius* of the circle.  Meaning that if we want a 10px-wide circle, then the want to have a 5px radius. And since the inner circle is a 6px-wide circle (with a radius of 3px), then the box-shadow `spread` property should be 2px to add up to a total radius of 5px, giving us a 10px-wide outer circle.

Enriqe commented:
This is ready for another review 👍

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -243,16 +252,56 @@
+  border-radius: 5px !important;
+}
+
+
```
Nit: line break

Enriqe replied:
Done.

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -243,16 +252,56 @@
+  align-items: center !important;
+  padding-top: 4px !important;
+  padding-bottom: 4px !important;
+  padding-right: 10px !important;
```
I imagine the compiler is smart enough to optimize this, but even for readability I think we could do `padding: 4px 10px 4px 0;`

Enriqe replied:
Done.

Enriqe commented:
> Would be good to add a visual test for this; could come as a separate PR 

Sounds good, will do in a follow up PR
